File: src/whatever2sbom/pipeline.py
import logging

from whatever2sbom.collectors.base import Collector
from whatever2sbom.enrichers.base import Enricher
from whatever2sbom.formatters.base import Formatter
from whatever2sbom.models import PackageRecord
from whatever2sbom.plugins.base import LoadedPlugin
from whatever2sbom.util import perf
from whatever2sbom.validators.base import ValidationError, Validator

logger = logging.getLogger(__name__)


class SbomPipeline:
    def __init__(
        self,
        collector: Collector,
        enrichers: list[Enricher],
        formatter: Formatter,
        validators: list[Validator],
        plugins: list[LoadedPlugin] | None = None,
    ) -> None:
        self.collector = collector
        self.enrichers = enrichers
        self.formatter = formatter
        self.validators = validators
        # Post-processing plugins run last, after formatting but *before*
        # validation, so whatever they emit is still checked against the schema.
        self.plugins = plugins or []

    def run(self) -> dict:
        logger.info("Collecting -> %s", self.collector.name)
        with perf.timed(f"collect:{self.collector.name}"):
            packages: list[PackageRecord] = self.collector.collect()

        for enricher in self.enrichers:
            logger.info("Enriching -> %s", enricher.name)
            with perf.timed(f"enrich:{enricher.name}"):
                packages = enricher.enrich(packages)

        logger.info("Formatting -> %s", self.formatter.name)
        with perf.timed(f"format:{self.formatter.name}"):
            bom = self.formatter.format(packages)

        for plugin in self.plugins:
            logger.info("Plugin -> %s", plugin.name)
            with perf.timed(f"plugin:{plugin.name}"):
                bom = plugin.run(bom)

        for validator in self.validators:
            logger.info("Validating -> %s", validator.name)
            with perf.timed(f"validate:{validator.name}"):
                errors = validator.validate(bom)
                if errors:
                    raise ValidationError(errors)

        return bom

File: internal/pipeline/pipeline.go
// Package pipeline runs collect → enrich → format → validate as one unit.
package pipeline

import (
	"fmt"
	"log/slog"

	"whatever2sbom/internal/collector"
	"whatever2sbom/internal/enricher"
	"whatever2sbom/internal/formatter"
	"whatever2sbom/internal/validator"
)

// ValidationError carries the per-violation messages produced by a Validator.
type ValidationError struct {
	Errors []string
}

// Error implements error.
func (e *ValidationError) Error() string {
	return fmt.Sprintf("%d validation error(s)", len(e.Errors))
}

// Pipeline wires the stages together. Construct it via New and call Run once.
type Pipeline struct {
	Collector  collector.Collector
	Enrichers  []enricher.Enricher
	Formatter  formatter.Formatter
	Validators []validator.Validator
}

// New returns a configured pipeline. All stages run unconditionally.
func New(
	c collector.Collector,
	es []enricher.Enricher,
	f formatter.Formatter,
	vs []validator.Validator,
) *Pipeline {
	return &Pipeline{Collector: c, Enrichers: es, Formatter: f, Validators: vs}
}

// Run executes the pipeline. Returns the formatted BOM and any error.
// Validation failures return *ValidationError so the caller can render
// per-violation messages.
func (p *Pipeline) Run() (any, error) {
	slog.Info("collecting", "source", p.Collector.Name())
	pkgs, err := p.Collector.Collect()
	if err != nil {
		return nil, err
	}
	slog.Info("collected", "packages", len(pkgs))

	for _, e := range p.Enrichers {
		slog.Info("enriching", "via", e.Name())
		pkgs, err = e.Enrich(pkgs)
		if err != nil {
			return nil, err
		}
	}

	slog.Info("formatting", "via", p.Formatter.Name())
	bom, err := p.Formatter.Format(pkgs)
	if err != nil {
		return nil, err
	}

	for _, v := range p.Validators {
		slog.Info("validating", "via", v.Name())
		if errs := v.Validate(bom); len(errs) > 0 {
			return nil, &ValidationError{Errors: errs}
		}
	}

	return bom, nil
}
